Xy calibration goodies2

vedina opened the pull request:
- test for laser zeroing verifies if the Si peaks is at the correct position
- additional method for peak matching using scipy linear_assignment
- CalibrationModel and XCalibrationComponent options to select the match_method 
- demo notebook

vedina commented:
@georgievgeorgi  @nicocopez if you could go through the demo notebook and tell 
- what is not clear
- what you want to change and wondering how , etc 
then I will add explanations (or additional code)

kerberizer commented on src/ramanchada2/io/HSDS.py:
```diff
@@ -67,18 +67,22 @@ def write_nexus(filename: str,
     pass
 
+def sanitize_key(key: str) -> str:
+    return ''.join(c if ord(c) < 128 else '_' for c in key)
```
Instead of `ord(c) < 128`, maybe use [`c.isascii()`](https://docs.python.org/3/library/stdtypes.html#str.isascii) or even `c.isalnum()` if the key will be only alphanumeric?

vedina replied:
isascii should be fine. 
there are some .spc with metadata keys with specific unicode that are breaking hdf5 writing

vedina replied:
we don't need alphanum only, keys can contain brackets, dots, etc

kerberizer replied:
If we know what the valid characters are, maybe test for them? Because currently we would also allow some non-printable characters, e.g. ^G, BELL.

kerberizer commented on src/ramanchada2/protocols/calibration.py:
```diff
@@ -582,20 +614,22 @@ def __init__(self, laser_wl: int):
-        calmodel.derive_model_x(
+        import ramanchada2 as rc2
+        import ramanchada2.misc.constants as rc2const
+        from ramanchada2.protocols.calibration import CalibrationModel
```
Why do we put these imports here, and why do we import the whole rc2?

vedina replied:
this is example in pydoc

kerberizer replied:
Ugh, my bad, thanks!

kerberizer commented on src/ramanchada2/misc/utils/argmin2d.py:
```diff
@@ -145,3 +146,162 @@ def match_peaks(spe_pos_dict, ref):
+    Modified cost function with an order preservation penalty and priority weighting.
+    - `order_weight` increases penalty for large differences in the x-axis values.
+    - `priority_weight` decreases the cost for higher values in the y-axis for set_b points.
+    """
```
It might be nice to follow the agreed upon docstring format: https://docs.python.org/3/library/stdtypes.html#str.isascii

kerberizer commented on src/ramanchada2/misc/utils/argmin2d.py:
```diff
@@ -145,3 +146,162 @@ def match_peaks(spe_pos_dict, ref):
+    >>> spectrum_b = {102: 12, 106: 22, 111: 16}
+    >>> match_peaks(spectrum_a, spectrum_b)
+
+    """
```
This seems more like numpy docstring syntax.

kerberizer reviewed (commented):
Have not looked into the domain specific code for obvious reasons.

georgievgeorgi commented on src/ramanchada2/misc/utils/argmin2d.py:
Would be better to have implementations not related to argmin2d in a separate file.

vedina replied:
Agree !

vedina replied:
Done

georgievgeorgi commented on src/ramanchada2/protocols/calibration.py:
the one-class-per-file rule will increase readability

vedina replied:
I thought Python is against that rule

georgievgeorgi commented on tests/protocols/test_calibrationmodel.py:
```diff
@@ -86,22 +101,48 @@ def setup_module():
+    find_kw["prominence"] = spe_test.y_noise_MAD() * setup_module.calmodel.prominence_coeff
+    cand = spe_test.find_peak_multipeak(**find_kw)
+    fit_kw = {}
+    fitres = spe_test.fit_peak_multimodel(profile=setup_module.si_profile, candidates=cand, **fit_kw, no_fit=False)
```
`**fit_kw` here makes no sense after `fit_kw = {}`

georgievgeorgi commented on tests/protocols/test_calibrationmodel.py:
```diff
@@ -86,22 +101,48 @@ def setup_module():
+
+    plt.grid()
     plt.tight_layout()
     plt.savefig("test_calmodel_{}.png".format("laser_zeroing"))
```
Do you really need to generate plots in the tests? They are slow and do not believe that they are going to be used somehow as part of the tests.

georgievgeorgi commented on src/ramanchada2/misc/utils/matchsets.py:
```diff
@@ -0,0 +1,227 @@
+    return order_penalty
+
+
+def cost_function(p1, p2, order_weight=1.0, priority_weight=.1):
```
It is not clear what `p1` and `p2` are. One can search in the source and see that these are supposed to be two element lists having positions and intensities. Having the code written this way is difficult to follow and makes it more difficult to focus on the algorithms.

georgievgeorgi commented on src/ramanchada2/misc/utils/matchsets.py:
```diff
@@ -0,0 +1,227 @@
+    - `priority_weight` decreases the cost for higher values in the y-axis for set_b points.
+    """
+    order_penalty = order_weight * abs(p1[0] - p2[0])
+    priority_bonus = priority_weight * p2[1]  # Rewards points in set_b with higher second dimension values
```
This is not always desirable. We may have a fake high-amplitude reference/peak

georgievgeorgi commented on src/ramanchada2/misc/utils/matchsets.py:
```diff
@@ -0,0 +1,227 @@
+
+    # Use the Hungarian algorithm to find the optimal assignment
+    try:
+        row_ind, col_ind = linear_sum_assignment(cost_matrix)
```
If we ignore the amplitude values `linear_sum_assignment` provides worse result than the already implemented `argmin2d`.

Consider this example:
```python
from scipy.optimize import linear_sum_assignment

import ramanchada2 as rc2
import numpy as np
import matplotlib.pyplot as plt

x = np.array([1, 1.5, 2, 2.5, 3, 6, 7, 7.1, 9, 10, 11, 12])
y = np.array([1, 2, 3, 4, 5, 6, 7, 8.])

lsa_x_idx, lsa_y_idx = np.array(linear_sum_assignment(np.abs(np.subtract.outer(x, y))))
gg_x_idx, gg_y_idx = rc2.misc.utils.find_closest_pairs_idx(x, y)

line_kw = dict(alpha=.5)
for i in x:
    plt.axvline(i, **line_kw)
    
for i in y:
    plt.axhline(i, **line_kw)
    
plt.plot(x[lsa_x_idx], y[lsa_y_idx], 'x', markersize=10, mew=3, label='linear_sum_assignmet')
plt.plot(x[gg_x_idx], y[gg_y_idx], '+', markersize=10, mew=3, label='gg argmin2d')
plt.legend(loc='lower right')
plt.xlabel('locations 1')
plt.ylabel('locations 2')
```

georgievgeorgi commented on src/ramanchada2/misc/utils/matchsets.py:
```diff
@@ -0,0 +1,227 @@
+    min_value = min(ref.values())
+    max_value = max(ref.values())
+    if len(ref.keys()) > 1:
+        normalized_ref = {key: (value - min_value) / (max_value - min_value) for key, value in ref.items()}
```
should not subtract `min_value` as this way the smallest peak is getting lost. min/max makes more sense in a spectrum, not individual peaks

vedina commented:
- Resolved the review comments
- refactoring of the protocols/calibration into (semi)separate classes
- optional plots in tests (disabled by default)
- fixed bug in match_peaks_cluster in interpretation of clustering results
- added options for match methods
- added options for interpolators - we now have RBFInterpolator, PchipInterpolator, CubicSpline . Of these, only PchipInterpolator guarantees monotonic calibration curve, and when interpolating.  But Ne peaks don't cover the entire range, so we need (some) extrapolation. Hence added extrapolation option.
- option what to do if the results of applying the calibration curve is nonmontonic - ignore, set nans, drop nans, raise error

@georgievgeorgi @kerberizer
